Src/VimCore/CommonOperations.fs

```diff
@@ -2387,19 +2427,385 @@ type internal CommonOperations
+            ()
+
+    /// Run the specified action for all selections
+    member x.RunForAllSelections action =
+
+        // Unless there are multiple selections just do the action once and
+        // return its result normally.
+        if x.IsMultiSelectionSupported then
+            let selectedSpans = x.SelectedSpans |> Seq.toList
+            if selectedSpans.Length > 1 then
+                x.RunForAllSelectionsCore action selectedSpans
+            else
+                action()
         else
-            TrackingPointUtil.GetPointInSnapshot point pointTrackingMode snapshot
-            |> OptionUtil.getOrDefault (SnapshotPoint(snapshot, min point.Position snapshot.Length))
+            action()
+
+    /// Run the specified action for the specified selected spans
+    member x.RunForAllSelectionsCore action selectedSpans =
+
+        // Get the current kind of visual mode, if any.
+        let visualModeKind =
+            _vimTextBuffer.ModeKind
+            |> VisualKind.OfModeKind
+
+        // Get any mode argument from the specified command result.
+        let getModeArgument result =
+            match result with
+            | CommandResult.Completed modeSwitch ->
+                match modeSwitch with
+                | ModeSwitch.SwitchModeWithArgument (_, modeArgument) ->
+                    Some modeArgument
+                | _ ->
+                    None
+            | _ ->
+                None
+
+        // Get any linked transaction from the specified command result.
+        let getLinkedTransaction result =
+            match getModeArgument result with
+            | Some (ModeArgument.InsertWithTransaction linkedTransaction) ->
+                Some linkedTransaction
+            | _ ->
+                None
+
+        // Get any visual selection from the specified command result.
+        let getVisualSelection result =
+            match getModeArgument result with
+            | Some (ModeArgument.InitialVisualSelection (visualSelection, _)) ->
+                _globalSettings.SelectionKind
+                |> visualSelection.GetPrimarySelectedSpan
+                |> Some
+            | _ ->
+                None
+
+        // Get any switch to visual mode from the specifed command result.
+        let getSwitchToVisualKind result =
+            match result with
+            | CommandResult.Completed modeSwitch ->
+                match modeSwitch with
+                | ModeSwitch.SwitchMode modeKind ->
+                    match modeKind with
+                    | ModeKind.VisualCharacter -> Some VisualKind.Character
+                    | ModeKind.VisualLine -> Some VisualKind.Line
+                    | ModeKind.VisualBlock -> Some VisualKind.Block
+                    | _ -> None
+                | _ ->
+                    None
+            | _ ->
+                None
+
+        // Create a linked undo transaction.
+        let createTransaction () =
+            let name = "MultiSelection"

```

Comment:
Low pri: Consider including the count of selections in the transaction name. The name appears in our log output, may help to know how many selections there were if we get user reports of bugs.